File: old/binary_search_range/search_insert_position/search_insert_position.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>

using namespace std;

int binarySearchInsertionPosition(vector<int>& arr, int val) {
    int low = 0;
    int high = arr.size() - 1;
    
    while (low <= high) {
        int middle = low + (high - low) / 2;
        if (arr[middle] < val) {
            low = middle + 1;
        } else if (arr[middle] > val) {
            high = middle - 1;
        } else {
            return middle;
        }
    }

    return low;
}

int main() {
    vector<int> input = {0, 2, 4, 6, 8, 10};
    for (int i: input) {
        cout << i << " ";
    }
    cout << endl;
    srand(time(0));
    int val = (rand() % 10) + 1;
    int insertOrFoundIndex = binarySearchInsertionPosition(input, val);
    cout << val << " can be found or inserted at index: " << insertOrFoundIndex << endl;
    return 0;
}

File: old/binary_search_range/search_insert_position/search_insert_position.py
import random


def binary_search(arr, val):
    low = 0
    high = len(arr) - 1

    while low <= high:
        middle = low + (high - low) // 2
        if arr[middle] < val:
            low = middle + 1
        elif arr[middle] > val:
            high = middle - 1
        else:
            return middle

    return low  # because after low <= high loop, low will be either in val or more than val


def main():
    arr = [1, 3, 5, 7, 9]
    val = random.randint(0, 10)
    found_or_insert_index = binary_search(arr, val)
    print(
        f"{val} can be found or inserted at index {found_or_insert_index} in the array: {arr}"
    )


main()
